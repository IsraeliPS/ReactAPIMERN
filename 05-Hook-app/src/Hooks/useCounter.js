import { useState } from 'react';

export const useCounter = (initialState = 10) => {
  const [counter, setCounter] = useState(initialState);
  const increment = (value = 1) => {
    setCounter(counter + 1);
  };

  const decrement = (value = 1) => {
    if (counter === 0) return;
    setCounter(counter - 1);
  };

  const reset = () => {
    setCounter(initialState);
  };

  return {
    counter,
    reset,
    increment,
    decrement,
  };
};
