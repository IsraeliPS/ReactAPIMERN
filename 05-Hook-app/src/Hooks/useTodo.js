import { useEffect, useReducer } from 'react';

import { todoReducer } from '../components/08-useReducer/todoReducer';

const init = () => {
  return JSON.parse(window.localStorage.getItem('todos')) || [];
};

export const useTodo = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const todosCount = todos.length;
  const pendingTodosCount = todos.filter((todo) => !todo.done).length;

  useEffect(() => {
    window.localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    dispatch({
      type: 'delete',
      payload: todoId,
    });
  };

  const handleToggle = (todoId) => {
    dispatch({
      type: 'toggle',
      payload: todoId,
    });
  };

  const handleAddTodo = (newTodo) => {
    dispatch({
      type: 'add',
      payload: newTodo,
    });
  };

  return {
    handleAddTodo,
    handleToggle,
    handleDelete,
    todos,
    todosCount,
    pendingsTodosCount: pendingTodosCount,
  };
};
