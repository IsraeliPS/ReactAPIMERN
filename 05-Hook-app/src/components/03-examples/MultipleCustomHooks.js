import React from 'react'
import { useFetch } from '../../Hooks/useFetch'
import { useCounter } from '../../Hooks/useCounter'

import '../styles.css'

export const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1)
  const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`)
  const { author, quote } = !!data && data[0]

  //   console.log(author, quote)
  return (
    <div>
      <h1 className='text-center'> BreakingBad Quotes</h1>

      <hr />

      {
        loading
          ? (

            <div className='alert alert-info text-center'> Loading...</div>
            )
          : (

            <blockquote className='blockquote text-end'>
              <p className='mb-2'> {quote}</p>
              <footer className='blockquote-footer'>{author}</footer>
            </blockquote>
            )
      }
      <button
        className='btn btn-primary boton'
        onClick={increment}
      >
        Siguiente Quote
      </button>

    </div>
  )
}
