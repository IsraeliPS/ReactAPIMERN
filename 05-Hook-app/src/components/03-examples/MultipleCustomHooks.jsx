import React from 'react';
import { useFetch, useCounter } from '../../Hooks';

import '../styles.css';
import { LoadingQuote } from './LoadingQuote';
import { Quote } from './Quote';

export const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1);
  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );
  const { author, quote } = !!data && data[0];

  //   console.log(author, quote)
  return (
    <div>
      <h1 className='text-center'> BreakingBad Quotes</h1>

      <hr />

      {loading ? <LoadingQuote /> : <Quote author={author} quote={quote}/>}

      <button className='btn btn-primary boton' onClick={() => increment()}>
        Siguiente Quote
      </button>
    </div>
  );
};
