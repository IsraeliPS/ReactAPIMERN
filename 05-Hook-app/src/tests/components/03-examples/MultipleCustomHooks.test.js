import React from 'react'
import { shallow } from 'enzyme'
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks'
import { useFetch } from '../../../Hooks/useFetch'
import { useCounter } from '../../../Hooks/useCounter'
jest.mock('../../../Hooks/useFetch')
jest.mock('../../../Hooks/useCounter')

describe('Pruebas en <MultipleCustomHooks', () => {
  useCounter.mockReturnValue({
    counter: 10,
    increment: () => {}
  })

  test('Debe de mostrarse correctamente', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    })
    const wrapper = shallow(<MultipleCustomHooks />)
    expect(wrapper).toMatchSnapshot()
  })

  test('Debe mostrar la informacion', () => {
    useFetch.mockReturnValue({
      data: [{
        author: 'Israeli',
        quote: 'Probando el mock'
      }],
      loading: false,
      error: null
    })

    const wrapper = shallow(<MultipleCustomHooks />)

    expect(wrapper.find('.alert').exists()).toBe(false)
    expect(wrapper.find('.mb-2').text().trim()).toBe('Probando el mock')
    expect(wrapper.find('footer').text().trim()).toBe('Israeli')
  })
})
