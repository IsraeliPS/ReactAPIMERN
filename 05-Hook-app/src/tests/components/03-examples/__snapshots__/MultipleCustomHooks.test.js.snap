// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks Debe de mostrarse correctamente 1`] = `
<div>
  <h1
    className="text-center"
  >
     BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
     Loading...
  </div>
  <button
    className="btn btn-primary boton"
    onClick={[Function]}
  >
    Siguiente Quote
  </button>
</div>
`;
